internal/self_metrics/self_metrics.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package self_metrics
+
+import (
+	"context"
+	"fmt"
+
+	mexporter "github.com/GoogleCloudPlatform/opentelemetry-operations-go/exporter/metric"
+	metricapi "go.opentelemetry.io/otel/metric"
+	metricsdk "go.opentelemetry.io/otel/sdk/metric"
+	"github.com/GoogleCloudPlatform/ops-agent/confgenerator"
+	"go.opentelemetry.io/contrib/detectors/gcp"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric/instrument"
+	"go.opentelemetry.io/otel/sdk/metric/metricdata"
+	"go.opentelemetry.io/otel/sdk/resource"
+)
+
+var (
+	agentMetricsPrefixFormat = "agent.googleapis.com/%s"
+	formatter                = func(d metricdata.Metrics) string {
+		return fmt.Sprintf(agentMetricsPrefixFormat, d.Name)
+	}
+)
+
+type enabledReceivers struct {
+	metricsReceiverCountsByType map[string]int
+	logsReceiverCountsByType    map[string]int
+}
+
+func CountEnabledReceivers(uc *confgenerator.UnifiedConfig) (enabledReceivers, error) {
+	eR := enabledReceivers{
+		metricsReceiverCountsByType: make(map[string]int),
+		logsReceiverCountsByType:    make(map[string]int),
+	}
+
+	// Logging Pipelines
+	for _, p := range uc.Logging.Service.Pipelines {
+		for _, rID := range p.ReceiverIDs {
+			rType := uc.Logging.Receivers[rID].Type()
```

Comment:
Added the check to avoid `nil` reference error. 

Now the `for loop` for counting in Logging Pipelines and Metrics Pipelines share a lot of similar logic. Should I create another method `CountPipelineReceivers` to avoid the repetition (considering there may be cumbersome notation about the signature). WDYT @braydonk  ?

Reply:
Discussed offline. Applied proposed solution in https://github.com/GoogleCloudPlatform/ops-agent/pull/852/commits/bfcf6f120753878199ead5e7d746155a1542d89b using generics. Thank you!